### ISSUE TITLE
Fastlane Evolution

### ISSUE THREAD

**Issue by mollyIV** [fastlane/fastlane#14253]
`Fastlane` is an amazing tool that helps developers to automate a lot of tasks, saves many hours of developers work and handles boring stuff for them. Thank you `Fastlane` ❤️ 

That being said, I am very curios about the future of the tool. I've been looking at the opened issues to figure it out, however there is not many **enhancements** (9 issues) or **features** (4 issues) labeled issues. There is a few `swift` issues as well. Based on that it's quite hard to see what is the overall direction / vision the tool is heading to. 

So basically, a few questions came to my mind 😊

**1. Is there any document / forum / Slack or any other channel of communication that people could follow the next milestone / goal for Fastlane?**

It's always great to get a sneak peak of exciting stuff coming and it adds extra dose of excitement for contributors 😅 I really like the way [Swift Evolution](https://github.com/apple/swift-evolution) handles it 👏 

**2. Swift support**

There's [Fastlane.swift beta](https://docs.fastlane.tools/getting-started/ios/fastlane-swift) 🎊 already. The next big thing seems to be plugin support:

> Currently, Fastlane.swift does not have support for plugins. This is a work in progress and we will continue to update this doc with the current working condition of each feature as we move from beta to general availability.

Is there any plan to migrate some Fastlane tools to Swift in the future? 👍Why Fastlane decided to support Swift in the first place? 🤔 

So sorry, if that's a wrong place to ask this type of questions, although didn't know where I could ask them 🙈If so, don't hesitate to close this issue. 

Thank you for your time! 🙇🏻‍♂️

**Comment by janpio** [comment]
> there is not many **enhancements** (9 issues) or **features** (4 issues) labeled issues.

That is simply because we don't use those labels that much. I mostly label the issues by "code area" (which effectively are the legacy tool names that correspond to folders in the codebase) so it is easier to work on several things in the same area of the code in batches, and so that pattern become visible.

Fastlane is also a pretty mature tool, lots of things are already implemented or available via plugins. There is not much functionality left to add, and if there is, it often is to niche to find the support of someone that could write the Ruby code for it.

> Based on that it's quite hard to see what is the overall direction / vision the tool is heading to.

Right now most of the development is either 1. pain/bug/issue or 2. contributor driven.

1 effectively means that contributors like you and me just look at what is present here in the issues. If something is broken, people write about it and we try to help them and they or we implement a solution.

2 is whatever a contributor wants to see in fastlane. I am working on Windows support for example, as I use Window as my main machine every day. Swift was implemented for similar reasons, a contributor just wanted to be able to write their fastlane configuration and lanes in Swift (afaik).

So if you want fastlane to do a specific thing, your best bet is to open an issue, see if there is interest (and agreement that such a feature would be merged), then implement it yourself or get someone else to implement it.

**Comment by mollyIV** [comment]
Thank you @janpio 🙇 Your answer gave me a context I was missing. 
> There is not much functionality left to add, and if there is, it often is to niche to find the support of someone that could write the Ruby code for it.

So you think there's no many Ruby developers left 🤔 Back in the day Ruby was a first choice for automation tools, i.e. CocoaPods to not looking far away. Wondering what would've been chosen today for Fastlane? Another question is if Fastlane might port Ruby code to other language in the future? 🤔 (even though right now it might seem like a crazy idea).

**Comment by janpio** [comment]
Nah, you misunderstood me here. Not ruby is to niche, but the problems that are not solved by fastlane. 

If the usecase for a feature is literally just 1 person, it will be hard to find some contributor (like you or me) that wants to implement it and then commit to supporting the feature for the next weeks, months or years.

When there are big changes like with force 2FA or a new API from Apple, we usually find enough contributors (like you and me!) that want and can work on it.

**Comment by joshdholtz** [comment]
@mollyIV Thanks for bring this up! This is definitely a topic that isn't very transparent to the fastlane community 😊 @janpio's answers were pretty on point there. _fastlane_ is a pretty mature tool at this point. It interacts most of the Apple APIs that it currently can as well as most of the Google Play APIs that it currently can. We also offer the third party plugin system for anybody to add features they want to add to fastlane without going through the _fastlane_ PR process which also helps limit the amount of knowledge that maintainers need to know 😇

That's being said... I think each of the individual contributors have a personal list of things they would like to see added to or changed in _fastlane_. Those usually come through as WIP PRs that prototype what we want changed. @janpio's Window's support (as he mentioned) is one of the bigger ones and we greatly appreciate it 😍 I have goals to clean up action/plugin configuration option interface and add some improved usage around lane options. One of our other combined team goals is to eventually switch over to the new App Store Connect API (but it currently does not offer everything that we need). _fastlane_ at this point is pretty big and each of the core contributors has the areas of _fastlane_ in which they are experts in so that is where you will tend to see new features appear more often.

>2. Swift support

Swift support was added as something that was in beta as it was more/less a prototype to see the possibilities of adding new language interfaces on top of _fastlane_ as well as seeing the demand that _Fastlane.swift_ would generate without rewriting all of _fastlane_ in a different language.

_Fastlane.swift_ was a success from the standpoint that its a relatively small amount of code that generates the Swift interface and also there are relatively few issues that arise from it. It also was developed in a way where it fully leverages the Ruby implementation of _fastlane_ and actually ended up enhancing the Ruby action option interface by cleaning up and explicitly specifying the different types of options that can be passed into actions. After this thought it was kind of left alone as it was working. 

The hard part about making plugins work with `Fastlane.swift` is that the plugins Swift interface would need to be generated a little more dynamically than the actions are (since the actions get generated during a _fastlane_ release. I would totally love adding plugin support but it sadly comes lower in priority to most of the other issues that come in. We would be happy to see this being worked on by someone 😊

There is no vision right now to migrate any of the codebase of _fastlane_ into Swift. Ruby is a pretty good fit for _fastlane_ right now as it runs nicely on any platform, its a scripting language which fits the `Fastfile`, and also because of the enormous amount of gems that are available for us to make use of for actions and plugins. Not to say that some of these don't exist on Swift for us to use but at the time when _fastlane_ was first being developed, it was much easier to get running on Ruby.

Rewriting the whole _fastlane_ codebase in Swift (or a different language) wouldn't provide much benefit for the amount of work it would take to replicate everything. The reason was Swift chosen as a second interface for _fastlane_ was that is would be a natural language for new iOS developers to write code in and that it was fairly straight forward for our team to implement as an interface because of iOS experience with Swift 🙃

### TL;DR
Sooooooo, this seemed to turn more into a blog post than a response to an issue 🙈
1. There is currently no one, single vision for _fastlane_'s future in terms of new features but we each have individual goals for _fastlane_ within our areas of expertise.
2. Swift was the more natural language for new iOS developers who want to adopt _fastlane_
